linter.py

```diff
@@ -1,57 +1,59 @@
 #
 # linter.py
-# Linter for SublimeLinter3, a code checking framework for Sublime Text 3
+# Linter for SublimeLinter4, a code checking framework for Sublime Text 3
 #
 # Written by Noah Davis
 # Copyright (c) 2016 Noah Davis
 #
+# Updated by Schuyler Jager & Andreas for SublimeLinter 4
+# Copyright (c) 2020 Schuyler Jager & Andreas
 # License: MIT
 #
 
 """This module exports the Codeclimate plugin class."""
 
-from SublimeLinter.lint import Linter, util, persist
+import re
+from SublimeLinter.lint import Linter, util
 
 
 class Codeclimate(Linter):
     """Provides an interface to codeclimate."""
 
-    syntax = (
-        'ruby on rails',
-        'ruby',
-        'javascript',
-        'python',
-        'go',
-        'html',
-        'css',
-        'markdown',
-        'php',
-        'haskall',
-        'java',
-        'shellscript'
-    )
-    executable = "codeclimate"
+    defaults = {
+        'chdir': "${project}",
+        'executable': 'codeclimate',
+        'selector': (
+            'source.css, '
+            'source.go, '
+            'source.haskall, '
+            'source.java, '
+            'source.javascript, '
+            'source.js, '
+            'source.php, '
+            'source.python, '
+            'source.ruby, '
+            'source.scss, '
+            'source.shell, '
+            'text.html'
+        )
+    }
     regex = r'^(?P<line>\d+)(?:-\d+)?:\s(?P<message>.+)$'
     multiline = False
     line_col_base = (1, 1)
     tempfile_suffix = '-'
     error_stream = util.STREAM_BOTH
-    selectors = {}
     word_re = None
-    defaults = {
-        'chdir': "${project}"
-    }
-    inline_settings = None
-    inline_overrides = None
-    comment_re = None
+
+    root = None
+    path = None
 
     def cmd(self):
-        """Construct a cmd to provide a relative path to 'codeclimate analyze'."""
-        result = [self.executable_path, 'analyze']
-        relative_file_name = None
-        for folder in self.view.window().folders():
-            if self.filename.find(folder) == 0:
-                relative_file_name = self.filename.replace(folder + '/', '')
-        result.append(relative_file_name)
-        persist.debug(result)
-        return result
+        """Set working directory and run 'codeclimate analyze'."""
+        if self.context.get('project_root') is None:
+            self.defaults['chdir'] = '${folder}' \
+                if self.context.get('folder') is not None \
+                else '${file_path}'
+
+        root = re.search(r"\{(\w+)\}", self.defaults['chdir']).group(1)
+        path = self.filename.replace(self.context.get(root) + '/', '')
+        return ['codeclimate', 'analyze', path]
```

Comment:
So generally, you must or want to set the *relative* path to the current file. *Relative* to what?  Usually the `working_dir`, no?  (I don't see this requirement in the codeclimate docs btw.)  

You can just get the working dir using `self.get_working_dir()` and compute the relative path to the filename, for example using `os.path.relpath(self.filename, self.get_working_dir())` (add `import os` at the top of the file) as it is Windows compatible.  (Although it looks like codeclimate is not either compatible or as easy to use on Windows.)  As long as it's a file-only-linter this could work as you don't have to expect `None` values for `filename` or `cwd`.  You do have to expect that the user opens random files within one window though.  So a check that `filename` is actually under control of the `folder` seems reasonable.

Generally ` return  ["codeclimate", "analyze", "$args", path]` t.i. somehow telling *where* user `args` should be injected is usually a good idea.  

Generally, you probably don't want to read in the `defaults` (`self.defaults[...]`).  You do want to read the actual, final settings using `self.settings[...]`.  Otherwise you don't get the values the user specified but always the defaults.  *Setting* (mutating) `self.defaults` is probably not what you want for the same reasons.

Reply:
Thank you very much for your review @kaste.

Your explanations helped me a lot to better understand how SublimeLinter works. I tried to improve my work based on your comments. Please feel free to request further changes.